### ISSUE TITLE
Generate Adaptive Theme From a Set of Basic Colors

### ISSUE THREAD

**Issue by hyrious** [hyrious/hyrious#33]
Sublime Text has a built-in theme called ["Adaptive"](https://www.sublimetext.com/blog/articles/the-adaptive-theme) which adapts to the syntax highlight theme. It uses some techniques which are very new in today's CSS many years ago, namely the [`color()` mod function](https://www.sublimetext.com/docs/minihtml.html#colors). Let's see some examples and try to represent them in CSS using [relative colors](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_colors/Relative_colors).

```json
"medium_dark_bg": "color(var(--background) blend(black 60%))",
"link_fg": "hsl(215, 60%, 50%)",
"vcs_modified": "color(var(--bluish) min-contrast(var(--background) 2.5))",
"icon_light_tint": "color(black a(0.7))",
"kind_name_label_border_color": "color(var(--accent) a(0.8))",
```

Think of the theme file as a function which accepts the syntax highlighting file (i.e. color scheme file) and outputs all colors applying to the whole app. It firstly extracts the interesting part from the color scheme, which is _colors_. See [predefined variables](https://www.sublimetext.com/docs/minihtml.html#predefined-variables). The `redish` color means the color closest to the basic `red`, and so on. A basic classification can be made to sort current color scheme into [dark | medium_dark | medium | light](https://www.sublimetext.com/docs/themes.html#luminosity) categories (which are effectively the "dark_high_contrast | dark | light | light_high_contrast" modes in GitHub and VS Code).

The next step is using color mod functions and some additional colors to define the whole theme. For example "In 'medium_dark' themes, let the whole app's background be the code's background plus 60% black", results in the definition of

```json
"medium_dark_bg": "color(var(--background) blend(black 60%))",
```

The equal CSS could be

```css
--dark-bg: hsl(from var(--background) h s calc(l * 0.6));
```

Likewise, we can only modify the alpha part of the color

```css
--kind-name-label-border-color: color(from var(--accent) srgb r g b / 0.8)
```

However, currently there's no `-ish` like things in CSS (find the nearest color a set of colors). I believe there should be a PostCSS plugin doing this thing but haven't find one. On the other hand, it is more likely to be predefined before generating the theme instead of calculated in theme itself. Anyway, finding the nearest color in colors [is not hard](https://github.com/dtao/nearest-color).

For `min-contrast()`, there's a thing in [CSS 6](https://drafts.csswg.org/css-color-6/#colorcontrast) called `contrast-color(bg vs fg1, fg2, ..., fallback)` which selects out an `fg` which has the highest contrast against `bg`. They are different things. I can come up with a rough implementation for `min-contrast(fg, bg, ratio)`, that it changes the `lightness` part of `fg` until it matches the ratio.

In summary, it might be more easily to generate app's theme from only a small set of colors, including `bg` `fg` `accent` `redish` `orangish` `yellowish` `greenish` `cyanish` `bluish` `purplish` and `pinkish`. The `accent` can be seen as the "primary" color to be used in links, buttons and outlines.

**Comment by hyrious** [comment]
Just found one thing that seems a disadvantage of using "modern" CSS features, especially `oklch()` (_lightness_, chroma and hue). Because modern displayers do have HDR mode and some of them could provide a _lightness_ range from 1 nit to 1600 nit (some macOS devices). But there're many other users still using 10 nit to 100 nit or even lower.

That will result in a very funny situration: some parts of the site get "brighter" on some devices than other devices. The contrast model (WACG) is broken here and we have no way to fix. (The actuall contrast feeling differs on different devices and settings, but color-p3 makes it even worse.)

So instead of using `oklch()` to make ["harmony"](https://github.com/evilmartians/harmony) color palettes. I have an idea that we keep using the old-school standard values (hex and hsl) to define and use colors.

Modern CSS features are still there for use, for example relative colors and `color-mix()` both support running under the `srgb` color space.
